### DISCUSSION TITLE
AxiosHttp with Lists doesn't work

### DISCUSSION THREAD

**Discussion by APTy** [lowdefy/lowdefy#655] [Questions and Help] [answered]
**Describe the bug**
Using the `AxiosHttp` Connection/Request with Lists doesn't work.

**To Reproduce**
Clone this fork of the `lowdefy-example-crud` that uses the Axios connection instead of the MongoDB one: https://github.com/apty/lowdefy-example-crud.

Run `npx lowdefy dev`.

Attempt v1: The page doesn't load at all, and the `lowdefy.contexts["brands:brands:{}"].state.brands` object is **empty** (even though the XHR in network tab succeeds).

Attempt v2: Changing `areas.content[].id` to `brands_list` (different from `brands` which is the name of the state var), allows the page to load with correct data in `lowdefy.contexts["brands:brands:{}"].state.brands`, but the UI cards don't load in that case. (see https://github.com/APTy/lowdefy-example-crud/commit/0bd5ce3bc88865177f14350f8da3a071a77f6800#diff-005cfa5022a2566b1a014cc6c21acf35f6137c6f8bfe229c71b7d7360af7f3ecR56)

**Expected behavior**
I expected the data loaded from the Axios connection to show up in the UI.

**Screenshots**
**Attempt v1 - Page Doesn't Load**
<img width="1789" alt="Screen Shot 2021-06-23 at 11 11 44 AM" src="https://user-images.githubusercontent.com/9010671/123122689-39dab100-d414-11eb-8795-ad6125d82958.png">

**Attempt v2 - Page loads but no data shows up**
<img width="1790" alt="Screen Shot 2021-06-23 at 11 13 47 AM" src="https://user-images.githubusercontent.com/9010671/123122671-37785700-d414-11eb-84a1-6597947aea7f.png">

**Lowdefy Version**
```
3.18.0
```

**Other Notes**
It'd be nice if there were more console logging about the issues being encountered. The `lowdefy` console var is super helpful, but I don't have a lot of insight into what else is happening.

**Comment by SamTolmay** [comment, marked as answer]
Thanks for logging the issue.

There are two problems involved here. The one related to the first request is because of a duplicated block id (both the page id and the `List` block ids are `brands`). This didn't cause a crash on the old version the example was running. I have updated the example for v3.18.0, and fixed this. The duplicate id problem should be resolved by #529. This will cause the build to fail, with a proper error message.

The second is related to how lists work. A list block requires an array value in state that matches its block id. The list block will then render its content area for each item in that array block. So when you changed the `List` block's id to `brands_list`, you should also change the `set_brands` action to set to `brands_list` instead of `brands`. Additionally, since you are using an Axios request, the data returned is an object that includes the response headers and body, so the action should become:

```yaml
id: set_brands
type: SetState
params:
  brands_list:
    _request: get_brands.data # This might be 'get_brands.data.brands' or something, depending on the API response
```

Please let me know if anything is unclear

**Reply by APTy** [reply]
@SamTolmay thanks! This makes a lot of sense now, and I've got it working! A few notes:
- the `get_brands.data` thing is very useful and probably worth documenting somewhere
- there's also this `container` key in `brands.$.container` that seems to be a "special" key that doesn't exist on the data, is that right or did i miss some part of the mongodb data?
- also the `A list block requires an array value in state that matches its block id.` is super crucial and i had this _feeling_ it was the case but didn't see it explicitly spelled out anywhere

Anyway, thanks again for all the help!

**Reply by SamTolmay** [reply]
Awesome! Relating to your comments:
- The dot-notation on requests is a new feature that we added in v3.18, so we can add some more examples.
- `brands_list.$.container` is the id of the container block (Box) that is inside the list. Each block in the list needs to have a `.$` in it's id, that will be substituted with the array index. Since the container isn't an input block, this is just an id, and doesn't really have anything to do with the value in state.
- I updated the comments in the example to explain this relation a little more clearly. I will also work on the [Lists](https://docs.lowdefy.com/lists) section in the docs to try explain things a little more clearly.
